Src/Plugins/Security/SEC101_048.CloudantConnectionStringValidator.cs

```diff
@@ -0,0 +1,117 @@
+                                                ref fingerprint,
+                                                ref message);
+        }
+
+        protected override string IsValidStaticHelper(ref string matchedPattern,
+                                                      ref Dictionary<string, string> groups,
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            // We need uri and neither account nor password, or uri and both account and password.  Use XOR
+            if (!groups.TryGetNonEmptyValue("uri", out string uri) ||
+                (groups.TryGetNonEmptyValue("account", out string account) ^
+                groups.TryGetNonEmptyValue("password", out string password)))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint()
+            {
+                Uri = uri,
+                Account = account,
+                Password = password,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            // TODO: Create a unit test for this. https://github.com/microsoft/sarif-pattern-matcher/issues/258
+
+            var fingerprint = new Fingerprint(fingerprintText);
+            string uri = fingerprint.Uri;
+            string account = fingerprint.Account;
+            string password = fingerprint.Password;
+
+            // At this point account and password must be either both full or both empty.  Only check one
+            if (string.IsNullOrWhiteSpace(account))
+            {
+                using (HttpClient client = CreateHttpClient())
+                {
+                    HttpResponseMessage response = client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead).GetAwaiter().GetResult();
+                    if (response.StatusCode == HttpStatusCode.OK)
+                    {
+                        return ReturnAuthorizedAccess(ref message, uri);
+                    }
+
+                    message = CreateUnexpectedResponseCodeMessage(response.StatusCode, uri);
+                }
+            }
+            else
+            {
+                HttpClientHandler handler = new HttpClientHandler();
+                handler.Credentials = new NetworkCredential(account, password);
+                using (HttpClient client = new HttpClient(handler)
+                {
+                    BaseAddress = new Uri(uri),
+                })
+                {
+                    HttpResponseMessage response = client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead).GetAwaiter().GetResult();
```

Comment:
>HttpResponseMessage [](start = 20, length = 19)

dispose

Reply:
Done

---
In reply to: [586923935](https://github.com/microsoft/sarif-pattern-matcher/pull/238#discussion_r586923935) [](ancestors = 586923935)